pkg/cmd/roachtest/tests/decommission.go

```diff
@@ -993,6 +1006,94 @@ func runDecommissionRandomized(ctx context.Context, t test.Test, c cluster.Clust
 	}
 }
 
+// runDecommissionDrains tests that a decommission implies a drain.
+// This means that SQL connections of a decommissioning node will be closed near the end of decommissioning.
+// The test cluster contains 4 nodes and the fourth node is decommissioned.
+// While the decommissioning node has open SQL connections, queries should never fail.
+func runDecommissionDrains(ctx context.Context, t test.Test, c cluster.Cluster) {
+	var (
+		numNodes     = 4
+		pinnedNodeID = 1
+		decommNodeID = numNodes
+		decommNode   = c.Node(decommNodeID)
+	)
+
+	err := c.PutE(ctx, t.L(), t.Cockroach(), "./cockroach", c.All())
+	require.NoError(t, err)
+
+	c.Start(ctx, t.L(), option.DefaultStartOpts(), install.MakeClusterSettings(), c.All())
+
+	h := newDecommTestHelper(t, c)
+
+	run := func(db *gosql.DB, query string) error {
+		_, err = db.ExecContext(ctx, query)
+		t.L().Printf("run: %s\n", query)
+		return err
+	}
+
+	{
+		db := c.Conn(ctx, t.L(), pinnedNodeID)
+		defer db.Close()
+
+		// Increase the speed of decommissioning.
+		err = run(db, `SET CLUSTER SETTING kv.snapshot_rebalance.max_rate='2GiB'`)
+		require.NoError(t, err)
+		err = run(db, `SET CLUSTER SETTING kv.snapshot_recovery.max_rate='2GiB'`)
+		require.NoError(t, err)
+
+		// Wait for initial up-replication.
+		WaitFor3XReplication(t, db)
+	}
+
+	// Connect to node 4 (the target node of the decommission).
+	db := c.Conn(ctx, t.L(), decommNodeID)
+	defer db.Close()
+
+	// Decommission node 4 and poll its status during the decommission.
+	var (
+		maxAttempts = 50
+		retryOpts   = retry.Options{
+			InitialBackoff: time.Second,
+			MaxBackoff:     5 * time.Second,
+			Multiplier:     2,
+		}
+		// The expected output of decommission right before the node is marked as "decommissioned."
+		expDecommissioning = [][]string{
+			decommissionHeader,
+			{strconv.Itoa(decommNodeID), "true|false", "0", "true", "decommissioning", "false"},
+			decommissionFooter,
+		}
+		// The expected output of decommission once the node is finally marked as "decommissioned."
+		expDecommissioned = [][]string{
+			decommissionHeader,
+			{strconv.Itoa(decommNodeID), "true|false", "0", "true", "decommissioned", "false"},
+			decommissionFooter,
+		}
+	)
+	t.Status(fmt.Sprintf("decommissioning node %d", decommNodeID))
+	err = retry.WithMaxAttempts(ctx, retryOpts, maxAttempts, func() error {
+		o, e := h.decommission(ctx, decommNode, pinnedNodeID, "--wait=none", "--format=csv")
+		require.NoError(t, errors.Wrapf(e, "decommission failed: %v"))
+
+		// Near the end of the decommission, check if the SQL connection is still open.
```

Comment:
To simplify this code block a bit, how about:
```go
if err := cli.MatchCSV(o, expDecommissioning); err != nil {
  return err
}

// Still not done decommissioning; check to see if we're drained.
if drained := db.Ping() == nil; drained {
    return nil
}

// We haven't been decommissioned or drained, make sure SQL connections are functional.
queryErr := run(db, `SHOW DATABASES`)
require.NoError(t, queryErr)
return nil
```

I'm not fully following what the 5s retry achieves. Mind explaining?

Reply:
👍 Used your code block as inspiration to simplify it a bit. It doesn't quite work as it but made changes so it does.